Breaking changes with PCU

flagdanger opened the pull request:
Breaking changes with PCU:

- Ported PCU interface to not use global state
  - i.e., mesh creation requires passing in a PCU instance
- The new interface uses C++
- A C interface for PCU still exists in pcu_c.cc
  - Create an instance of the C handle struct "PCU_t"
  - The original pcu.c functions can be used by passing this struct
- Renamed source and headers files: 
  - PCU.h became PCU_C.h: PCU.h was repurposed as the C++ header
  - pcu.c became pcu_c.cc: pcu.c was deleted and pcu_c.cc created without history
  - more details on the history is here: https://github.com/SCOREC/core/pull/433#issuecomment-2455575371
- Build documentation (`make doc`) to generate doxygen based docs or see https://www.scorec.rpi.edu/pumi/doxygen/pcu.html for basic PCU.cc usage

flagdanger commented on parma/group/parma_group.cc:
```diff
@@ -27,25 +28,41 @@ typedef Parma_GroupCode GroupCode;
+  MPI_Comm_split(expandedPCU->GetMPIComm(), group, groupRank, &groupComm);
+  auto groupedPCU = std::unique_ptr<pcu::PCU>(new pcu::PCU(groupComm));
+  if (m){
+    m->switchPCU(groupedPCU.get());
```
switchPCU function may need review for a possible alternative method.
Also used in:
test/fusion.cc
test/ptnParma.cc
test/split.cc
test/ugrid.cc
test/zsplit.cc

jacobmerson commented:
@cwsmith This draft pull request is our starting point for having breaking changes of core with the new PCU as discussed in Pull request #388  .  The biggest thing that I don't like is the way parama group works because it still requires doing the comm switch operations.

However, my concern is that breaking that particular code will be a much more substantial thing that we want to force ourselves or the users deal with.

We are certainly open to suggestions on if there is a way out of that comm switch without totally blowing things up.

cwsmith commented:
@jacobmerson @flagdanger Can we meet next week to briefly discuss?  I see the call to switch in `test/ptnParma.cc` but would like to get the bigger picture view of the issue before digging into this more. I'll send an email.

cwsmith reviewed (commented):
I appreciate the effort taken to minimize the changes here.  I'm sure it was tempting to 'clean up'.

Breaking the dependency on the PCU global singleton is a big undertaking.  Thank you.  Excluding the C or C like C++ code, I'm wondering if there is a way to better integrate the use of the PCU object within classes structs so it isn't passed around through APIs as much.  Could inheritance from the PCU object help?

Some more detailed comments are below.

Thank you.

- Is the new interface for `Add` and related functions (`Max`, `Min`, etc.) templated on the type being passed in. There is a comment in  
parma/diffMC/parma_vtxPtnWriter.cc about a possible overflow due to a lost explicit cast.  I marked one other spot (in phasta?) that may be problematic.  I'm sure I missed others.
- It looks like `PCU_Comm_<name>2(....)` APIs that take the `PCUHandle` are replacing the calls to the same calls from the 'old' `PCU.h` that didn't have the `2` appended.  Can we remove the `2`?
- There appear to be two headers now `PCU2.h` and `PCUObj.h` which respectively are the C and C++ interface headers.  Can we remove the `2` from the C header and make the C++ header `PCU.hpp`?
  - Along those lines, in the diff pcu.(h|c) was deleted and replaced with pcu2.(h|c).  It would be nice if the C interface changes were recorded in the diff as opposed to deleted file + a new file.
- At first glance, the naming `PCUHandle` and `PCUObj` are a bit confusing.  Could we rename the C interface object `PCU_t` or something like that?
- There were a handful of locations where the chain of derefrences needed to get to the PCU object was > 4.  This doesn't seem critical at this point, but I'd like to avoid that syntax if there is a low maintenance way to avoid it.
  - Maybe there are portions of the code (e.g. phasta?, pumi?) where a struct/class could keep a copy of the PCU pointer (reference counted?) to clean up the interface a bit.
  - I'm a bit concerned about the `pumi_*` interfaces changing... I'm not sure why exactly.  Since it has a singleton, 'pumi::instance()->mesh->getPCU()' can we hide the changes?
  - Is the overhead of calling `m->getPCU()->someMethod()` significant? I'm guessing that outside of a 'hot loop' it isn't, but when we test the parma group code on an adapt case we should also check the wall time. 
  - Could the mesh creation functions `makeMdsBox(...)` and other object creation functions default to creating/owning the pcu object created on COMM_WORLD? 
- When we next meet I'd like to go over the `parma_group.cc` changes again.  Along those lines, running a 16 or 32 part `condense` test and a mesh adaptation case that uses parma_group to reduce the process count during adaptation should be on our todo list.
- I don't recall seeing changes to Parma_ShrinkPartition.  I may have just missed it.
- Are there unit tests for the new PCU interface?
- I skipped over many of the `test/*.cc` files as the changes are mostly the same.

jacobmerson commented:
@cwsmith Thanks for the review! This is a bit of a painful one as we had to touch so many files. Flynn is currently working on running the DeltaWing case. And once we have that sorted we will update this pull request.

@flagdanger please take a look at Cameron's review. Once you have a chance to look it over we can discuss the best approach for the areas that require more than simple changes.

flagdanger commented:
@jacobmerson @cwsmith 
The DeltaWing case is running, and I'm fixing the small changes, Cameron, you've mentioned now.

flagdanger commented on mds/mdsCGNS.cc:
The PCU_Comm_<name>2 API's still exist as we were unsure if we wanted to overwrite the original PCU function naming immediately. These can be changed to get rid of the 2.

flagdanger commented:
@cwsmith Do you have a model to use to partition the deltaWing_500kMetric.smb mesh, so we can run your adaptPumi case on multiple processors?

cwsmith commented:
@flagdanger Try passing '.null' as the argument; we don't have the CAD model for this mesh.

flagdanger commented:
@cwsmith @jacobmerson Made changes based on Cameron's comments, I believe this is ready to be looked at again.

cwsmith commented on pumi/pumi_mesh.cc:
Nice cleanup on the pumi_* files.  Thank you.

cwsmith commented on pumi/pumi_mesh.cc:
```diff
@@ -205,21 +205,21 @@ pGeom pumi_mesh_getGeom(pMesh)
   if (strcmp(mesh_type,"mds"))
   {
-    if (!PCUObj->Self()) std::cout<<"[PUMI ERROR] "<<__func__<<" failed: invalid mesh type "<<mesh_type<<"\n";
+    if (!pumi::instance()->getPCU()->Self()) std::cout<<"[PUMI ERROR] "<<__func__<<" failed: invalid mesh type "<<mesh_type<<"\n";
```
note to self: long dereference chains here

cwsmith commented on pumi/pumi_geom.cc:
```diff
@@ -191,9 +191,9 @@ void pumi_giter_reset(gIter iter)
+void pumi_geom_print (pGeom g, bool print_ent)
 {
-  if (PCUObj->Self()) return;
+  if (pumi::instance()->getPCU()->Self()) return;
```
calling `pumi_rank()` would be cleaner

cwsmith commented on pumi/pumi_geom.cc:
```diff
@@ -73,11 +73,11 @@ pGeom pumi_geom_load(gmi_model* gm, pcu::PCU *PCUObj, const char* model_type,
   }
 
-  if (!PCUObj->Self() && filename)
+  if (!pumi::instance()->getPCU()->Self() && filename)
```
calling `pumi_rank()` would be cleaner

cwsmith commented on test/constructThenGhost.cc:
```diff
@@ -44,10 +45,10 @@ int main(int argc, char** argv)
   m->verify();
 
-  if (!pumi_rank(pcu_obj.get())) printf("model/mesh converted to pumi instance\n");
+  if (!pcu_obj.get()->Self()) printf("model/mesh converted to pumi instance\n");
```
can `pumi_rank()` be called here?

cwsmith reviewed (changes requested):
Looks good.  A few small changes are requested below.

There are some items from the previous review we should discuss.

cwsmith commented:
/runtests

cwsmith commented:
/runtests

cwsmith commented:
@flagdanger Would you please take a look at the the CI build failure with simmetrix enabled:

https://github.com/SCOREC/core/actions/runs/10118195635/job/27984611512#step:3:405

```
cd /lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/_temp/build/mds && /opt/scorec/spack/rhel9/v0201_4/install/linux-rhel9-x86_64/gcc-12.3.0/mpich-4.1.1-xpoyz4tqgfxtrm6m7qq67q4ccp5pnlre/bin/mpicc -DHAVE_SIMMETRIX -DOMPI_SKIP_MPICXX -I/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/_temp/build/mds -I/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/pcu -I/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/pcu/reel -I/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/gmi -I/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/lion -I/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf -I/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/can -I/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/mth -g -Werror -Wall -Wextra -Wno-strict-overflow  -O3 -DNDEBUG -MD -MT mds/CMakeFiles/mds.dir/mds_apf.c.o -MF CMakeFiles/mds.dir/mds_apf.c.o.d -o CMakeFiles/mds.dir/mds_apf.c.o -c /lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/mds/mds_apf.c
/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf_sim/apfSIM.cc: In function 'apf::Mesh2* apf::createMesh(pParMesh, pcu::PCU)':
/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf_sim/apfSIM.cc:1108:31: error: cannot convert 'pcu::PCU' to 'pcu::PCU*'
 1108 |     m->init(getSerendipity(), PCUObj);
      |                               ^~~~~~
      |                               |
      |                               pcu::PCU
In file included from /lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf/apfMesh2.h:14,
                 from /lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf_sim/apfSIM.h:6,
                 from /lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf_sim/apfSIM.cc:1:
/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf/apfMesh.h:111:40: note:   initializing argument 2 of 'void apf::Mesh::init(apf::FieldShape*, pcu::PCU*)'
  111 |     void init(FieldShape* s, pcu::PCU *PCUObj);
      |                              ~~~~~~~~~~^~~~~~
/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf_sim/apfSIM.cc:1110:33: error: cannot convert 'pcu::PCU' to 'pcu::PCU*'
 1110 |     m->init(getLagrange(order), PCUObj);
      |                                 ^~~~~~
      |                                 |
      |                                 pcu::PCU
/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf/apfMesh.h:111:40: note:   initializing argument 2 of 'void apf::Mesh::init(apf::FieldShape*, pcu::PCU*)'
  111 |     void init(FieldShape* s, pcu::PCU *PCUObj);
      |                              ~~~~~~~~~~^~~~~~
```

The environment and cmake  commands used for the CI build on a SCOREC RHEL9 system are here:

https://github.com/SCOREC/core/blob/master/.github/workflows/simmetrix_enabled_pr_comment_trigger_self_hosted.yml#L34-L59

flagdanger commented:
> @flagdanger Would you please take a look at the the CI build failure with simmetrix enabled:
> 
> https://github.com/SCOREC/core/actions/runs/10118195635/job/27984611512#step:3:405
> 
> ```
> cd /lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/_temp/build/mds && /opt/scorec/spack/rhel9/v0201_4/install/linux-rhel9-x86_64/gcc-12.3.0/mpich-4.1.1-xpoyz4tqgfxtrm6m7qq67q4ccp5pnlre/bin/mpicc -DHAVE_SIMMETRIX -DOMPI_SKIP_MPICXX -I/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/_temp/build/mds -I/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/pcu -I/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/pcu/reel -I/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/gmi -I/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/lion -I/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf -I/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/can -I/lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/mth -g -Werror -Wall -Wextra -Wno-strict-overflow  -O3 -DNDEBUG -MD -MT mds/CMakeFiles/mds.dir/mds_apf.c.o -MF CMakeFiles/mds.dir/mds_apf.c.o.d -o CMakeFiles/mds.dir/mds_apf.c.o -c /lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/mds/mds_apf.c
> /lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf_sim/apfSIM.cc: In function 'apf::Mesh2* apf::createMesh(pParMesh, pcu::PCU)':
> /lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf_sim/apfSIM.cc:1108:31: error: cannot convert 'pcu::PCU' to 'pcu::PCU*'
>  1108 |     m->init(getSerendipity(), PCUObj);
>       |                               ^~~~~~
>       |                               |
>       |                               pcu::PCU
> In file included from /lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf/apfMesh2.h:14,
>                  from /lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf_sim/apfSIM.h:6,
>                  from /lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf_sim/apfSIM.cc:1:
> /lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf/apfMesh.h:111:40: note:   initializing argument 2 of 'void apf::Mesh::init(apf::FieldShape*, pcu::PCU*)'
>   111 |     void init(FieldShape* s, pcu::PCU *PCUObj);
>       |                              ~~~~~~~~~~^~~~~~
> /lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf_sim/apfSIM.cc:1110:33: error: cannot convert 'pcu::PCU' to 'pcu::PCU*'
>  1110 |     m->init(getLagrange(order), PCUObj);
>       |                                 ^~~~~~
>       |                                 |
>       |                                 pcu::PCU
> /lore/cwsmith/githubSelfHostedRunner/actions-runner-2.317.0/_work/core/core/core_433/apf/apfMesh.h:111:40: note:   initializing argument 2 of 'void apf::Mesh::init(apf::FieldShape*, pcu::PCU*)'
>   111 |     void init(FieldShape* s, pcu::PCU *PCUObj);
>       |                              ~~~~~~~~~~^~~~~~
> ```
> 
> The environment and cmake commands used for the CI build on a SCOREC RHEL9 system are here:
> 
> https://github.com/SCOREC/core/blob/master/.github/workflows/simmetrix_enabled_pr_comment_trigger_self_hosted.yml#L34-L59

Yes I'll look now, thanks.

flagdanger commented:
@cwsmith Fixed the things you mentioned, and can run tests again.

cwsmith commented:
/runtests

flagdanger commented:
@cwsmith Can you run again, it failed with this error:
This request was automatically failed because there were no enabled runners online to process the request for more than 1 days

cwsmith commented:
/runtests

flagdanger commented:
@cwsmith Thank you, could you run them again? And yes, if this fails again relating to simmetrix, then I will want to be able to reproduce these errors on my machine.

cwsmith commented:
/runtests

cwsmith commented:
It looks like the current errors are promoted warnings that are unrelated to this PR:

https://github.com/SCOREC/core/actions/runs/10269910071/job/28416451838#step:3:805

I'll see if I can resolve these quickly in develop.

flagdanger commented:
@cwsmith Okay I believe it is fixed.

cwsmith commented:
/runtests

flagdanger commented:
@cwsmith Okay can you run again, I'm not getting these errors on my machine.

cwsmith commented:
/runtests

flagdanger commented on test/generate.cc:
```diff
@@ -42,6 +42,7 @@
 #include "SimAttribute.h"
 #include "ModelTypes.h"
 
+static const pcu::PCU* globalPCU = nullptr;
```
Issue here with messageHandler() and lines: 64, 68, and 300. Temporary fix is a global variable, but will be looked at soon.

jacobmerson replied:
@cwsmith I don't think we remembered to discuss this on Friday. The Simmetrix message handler doesn't provide much of an option to get any data into it which requires a global variable.

cwsmith replied:
This is fine with me.

flagdanger commented:
@cwsmith Could you run tests again?

cwsmith commented:
/runtests

cwsmith commented on parma/group/parma_group.cc:
```diff
@@ -27,25 +28,41 @@ typedef Parma_GroupCode GroupCode;
   code.run(group);
-  PCU_Switch_Comm(oldComm);
+  
   MPI_Comm_free(&groupComm);
```
Yes; PCU duplicates the comm.

cwsmith commented on apf/apfMesh.cc:
```diff
@@ -5,7 +5,7 @@
  */
 
-#include <PCU.h>
+#include <PCUObj.h>
```
Renaming PCU.h to PCU_C.h and PCUObj.h to PCU.h as the C++ interface will be the primary interface for users/developers in C++.

cwsmith commented:
@cwsmith Todo:
- enable capstone CI build on self-hosted runner: https://github.com/SCOREC/core/issues/437

cwsmith commented on pcu/pcu_coll.c:
```diff
@@ -34,132 +33,33 @@ void pcu_merge_assign(void* local, void* incoming, size_t size)
 /* initiates non-blocking calls for this
    communication step */
-static void begin_coll_step(pcu_coll* c)
+static void begin_coll_step(pcu_mpi_t* mpi, pcu_coll* c)
```
`pcu_mpi_t` used to be global

cwsmith commented on phasta/phOutput.cc:
```diff
@@ -32,23 +31,23 @@ static void getCounts(Output& o)
   int numElms = o.mesh->count(dim);
-  long sumElms = PCU_Add_Long(numElms);
-  double avgElms = static_cast<double>(sumElms) / PCU_Comm_Peers();
+  long sumElms = o.mesh->getPCU()->Add(numElms);
```
change to `long sumElms = o.mesh->getPCU()->Add<long>(numElms);`

flagdanger commented on ma/maBalance.cc:
```diff
@@ -135,7 +134,7 @@ double estimateWeightedImbalance(Adapt* a)
 void preBalance(Adapt* a)
 {
-  if (PCU_Comm_Peers()==1)
+  if (a->mesh->getPCU()->Peers()==1)
```
@cwsmith @jacobmerson I got rid of all the >4 dereference chains, there only remain about 10 of these 4 long chains. Let me know if you'd like me to shorten them, although I think this format can sometimes make it easier to immediately see where the mesh is coming from.

jacobmerson replied:
I'm actually fine with the dereference chains. If Cameron prefers, just make a local variable with `PCU*` at the top of all of those functions.

flagdanger commented:
@cwsmith, I also changed the file naming of PCUObj.h and PCU.h,  and added explicitly declared data types to the templated add, min, max functions. I'm not seeing in the reviews any other changes we'd talked about making, but let me know if you'd like anything done differently.

jacobmerson commented on crv/crvShape.cc:
```diff
@@ -635,7 +634,7 @@ static int markEdgesOppLargeAnglesTri(Adapt* a)
     m->end(it);
   } while(count > prev_count);
-  return PCU_Add_Long(count);
+  return m->getPCU()->Add<int>(count);
```
should be `Add<long>`

jacobmerson commented on crv/crvShape.cc:
```diff
@@ -663,7 +662,7 @@ static int markEdgesOppLargeAnglesTet(Adapt* a)
     m->end(it);
   } while(count > prev_count);
-  return PCU_Add_Long(count);
+  return m->getPCU()->Add<int>(count);
```
`Add<long>`

jacobmerson commented on crv/crvShape.cc:
```diff
@@ -701,28 +700,28 @@ static int markEdgesToFix(Adapt* a, int flag)
   m->end(it);
 
-  return PCU_Add_Long(count);
+  return m->getPCU()->Add<int>(count);
```
`Add<long>`

jacobmerson commented on crv/crvShape.cc:
```diff
@@ -733,30 +732,30 @@ static void collapseInvalidEdges(Adapt* a)
-  double t1 = PCU_Time();
-  ma::print("Collapsed %d bad edges "
-      "in %f seconds",successCount, t1-t0);
+  successCount = m->getPCU()->Add<int>(successCount);
```
`Add<long>`

jacobmerson commented on apf/apfCGNS.cc:
```diff
@@ -635,7 +633,7 @@ void AddBocosToMainBase(const CGNS &cgns, const CellElementReturn &cellResults,
       const int number = bc.second.size();
       int total = number;
-      PCU_Add_Ints(&total, 1); // size of total array
+      m->getPCU()->Add<int<(&total, 1); // size of total array
```
should this be `Add<int>`?

jacobmerson commented on apf/apfConvertTags.h:
```diff
@@ -87,71 +86,71 @@ apf::MeshTag* setMappedTag(Mesh2* m, const char* tagName,
   T* c = new T[mySize*entries];
 
-  apf::Gid start = PCU_Exscan_Long(nverts);
+  apf::Gid start = m->getPCU()->Exscan(nverts);
```
`Exscan<long>`

jacobmerson commented on apf/apfMixedNumbering.cc:
```diff
@@ -228,8 +227,9 @@ static int number_ghost(
-  long start = PCU_Exscan_Long(long(dofs));
+    std::vector<GlobalNumbering*>& global,
+    pcu::PCU *PCUObj) {
+  long start = PCUObj->Exscan(long(dofs));
```
PCUObj->Exscan<long>(dofs);

cwsmith replied:
```suggestion
  long start = PCUObj->Exscan<long>(dofs);
```

jacobmerson replied:
Yup, Cameron is correct. I missed the type specification.

jacobmerson reviewed (changes requested):
Found at least some instances of mismatched types between the templated functions and the original ones.

cwsmith commented on pumi/pumi_mesh.cc:
```diff
@@ -69,10 +69,10 @@ void generate_globalid(pMesh m, pMeshTag tag, int dim, pOwnership o)
   m->end(it);
 
-  PCU_Exscan_Ints(&num_own,1);
+  m->getPCU()->Exscan(&num_own,1);
```
```suggestion
  m->getPCU()->Exscan<int>(&num_own,1);
```

cwsmith commented on test/matchedNodeElmReader.cc:
```diff
@@ -438,7 +438,7 @@ void getLocalRange(apf::Gid total, int& local,
     }
   }
-  first = PCU_Exscan_Long(local);
+  first = PCUObj->Exscan(local);
```
```suggestion
  first = PCUObj->Exscan<long>(local);
```

cwsmith commented on test/fusion2.cc:
```diff
@@ -92,7 +92,7 @@ struct GroupCode : public Parma_GroupCode
   void run(int group)
   {
-    mesh = ::makeEmptyMesh();
+    //mesh = ::makeEmptyMesh(PCUObj);
```
Should this be commented out?

cwsmith commented on test/fusion2.cc:
```diff
@@ -140,17 +140,19 @@ static void globalCode(apf::Mesh2* m)
+  Parma_SplitPartition(code.mesh, groupSize, code, PCUObj.get());
   /* update mesh for leaving groups */
-  apf::remapPartition(code.mesh, outMap);
+  //apf::remapPartition(code.mesh, outMap);
```
Should this be commented out?

cwsmith commented on test/ma_test_analytic_model.cc:
```diff
@@ -171,10 +171,11 @@ apf::Mesh2* createSphereMesh()
   MPI_Init(&argc,&argv);
-  PCU_Comm_Init();
+  {
+  auto PCUObj = std::unique_ptr<pcu::PCU>(new pcu::PCU(MPI_COMM_WORLD));
```
Can a PCU constructor/factory method be created to do this?

jacobmerson replied:
I think if we want to keep this as a `unique_ptr` `make_unique` should be used. Is there a reason to have a `unique_ptr` rather than just constructing a PCU object and passing the address to any of the core functions?

cwsmith replied:
If we could drop the unique_ptr that would be ideal; users will copy what is done in these examples and we should try to follow/present 'best practices'.

cwsmith reviewed (changes requested):
Looks good.  Thank you.

@flagdanger Do you have the output from the delta wing case?

Besides the comments/edits below, we need to capture a summary of the changes in the opening comment so it can be added to the merge commit.

jacobmerson commented on mds/apfMDS.cc:
```diff
@@ -195,11 +195,11 @@ class MeshMDS : public Mesh2
-      mesh = mds_read_smb(m, pathname, 0, this);
-      isMatched = PCU_Or(!mds_net_empty(&mesh->matches));
+      init(apf::getLagrange(1), PCUObj);
+      mesh = mds_read_smb(this->getPCU()->GetCHandle(), m, pathname, 0, this);
+      isMatched = this->getPCU()->Or(!mds_net_empty(&mesh->matches));
```
`this` pointer is not needed explicitly, but this has no impact on the semantics. The compiler will insert this implicitly for any member functions.

cwsmith commented:
@jacobmerson 
- The issue with the CGNS build of pumi on the github runner issue is still being resolved.  Hopefully I'll have an update on it tomorrow afternoon.  
- I've added you to the approved list of users who can trigger the tests via `runtests`.  That change should merge into master in the morning and you can try `runtests' again.

cwsmith commented:
Update:  
- The github hosted runner had its core count increased to eight to support the larger process counts the CGNS tests need. 
- HDF5 (needed by CGNS) has file locking requirements (see https://github.com/HDFGroup/hdf5/blob/develop/doc/file-locking.md) which `/lore` didn't appear to support.
- The github runner is now working out of `/space` to avoid this.

/runtests

cwsmith commented:
It looks like adding the C++14 standard into the mix for the cgns build causes some more errors to be emitted:
https://github.com/SCOREC/core/actions/runs/10604091991/job/29389889050#step:4:187

jacobmerson commented:
@flagdanger looks like that error is just an extra semicolon

cwsmith commented:
/runtests

cwsmith commented:
/runtests

flagdanger commented:
@cwsmith Could you run tests again?

cwsmith commented:
/runtests

jacobmerson commented:
@flagdanger @cwsmith Is all that's left fixing merge conflicts and rerunning the tests?

cwsmith reviewed (commented):
Looks good.  I saw only one left over `unique_ptr` use.

cwsmith commented on pcu/pcu_c.cc:
@flagdanger I'd like to maintain the linkage between these files so history is preserved.  Please let me know if you need help with that.

flagdanger replied:
Yes, I forgot. I might need help doing this, I was having trouble before.

cwsmith replied:
OK.  No problem.  I'll message you offline to setup a time.

cwsmith commented:
> @flagdanger @cwsmith Is all that's left fixing merge conflicts and rerunning the tests?

Yeah, and we need to have a summary of the changes in the opening comment so it can be recorded in the merge commit1.

flagdanger commented:
@cwsmith Hey, one of the recent develop commits deleted all these capstone_clis files making them conflicts. Should I also have them removed from this pull?

cwsmith commented:
> @cwsmith Hey, one of the recent develop commits deleted all these capstone_clis files making them conflicts. Should I also have them removed from this pull?

Good question.  Let's discuss during the meeting if we want to merge develop into this branch or rebase.

cwsmith commented:
todo:
- capstone files
- summary of the changes in the opening comment, and tests run
- check delta wing results in detail

cwsmith commented:
The history of `pcu.c` is complex and I don't think it is worth the effort to stitch the history back together.

### history of pcu.c in develop and the PCU API breaking branch

pcu.c renamed to pcu.cc in 2eada4eadad36a7e7b70173b003077448800b20f
```
commit 2eada4eadad36a7e7b70173b003077448800b20f
Author: Jacob Merson <jacob@jacobmerson.com>
Date:   Thu Mar 30 02:11:09 2023 -0400

    first pass of PCU with making use of a high level object

...
 pcu/{pcu.c => pcu.cc}  | 466 ++++++++++++++++++++++++++++++++++++++++---------------------------------
 pcu/pcu2.c             | 787 ----------------------------------------------------------------------------------------------------------------------------
...
 21 files changed, 798 insertions(+), 1201 deletions(-)
```

pcu.cc deleted in e33b48d52769a1835fae3198108eea4d4f5cf295
```
commit e33b48d52769a1835fae3198108eea4d4f5cf295
Author: flagdanger <supersags@gmail.com>
Date:   Mon Jun 3 13:28:38 2024 -0400

    removed last uses of old pcu, removed pcu.cc and .h, got rid of default
nullptr PCUObj arguments

...
 pcu/pcu.cc                                             | 581 --------------------------------------------------------------------------------------------
...
 35 files changed, 107 insertions(+), 1143 deletions(-)
```

### history of pcu_c.cc in the PCU API breaking branch

pcu2.cc created
```
commit 3c7b4479a3c320ff62e9ce17b78b0ec983f23696
Author: flagdanger <supersags@gmail.com>
Date:   Fri Jan 26 16:31:10 2024 -0500

    fixed last of test folder and made PCU2 .h and .cc 

...
 pcu/pcu2.cc                  | 518 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
...
 17 files changed, 691 insertions(+), 48 deletions(-)
```

pcu2.cc renamed to pcu.cc
```
commit 6e9a87525abfbd247bc2844f5cb79a3e7a64b886
Author: flagdanger <supersags@gmail.com>
Date:   Fri Jul 5 12:27:00 2024 -0400

    PCU2 using PCU naming, changed PCUHandle to PCU_t

...
 pcu/{pcu2.cc => pcu.cc} | 129 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-----------------------------------------------------------------
...
 24 files changed, 395 insertions(+), 406 deletions(-)
```

pcu.cc renamed to pcu_c.cc 
```
commit 7025e6214b0672a70b4a429c62fe21dc4cd1ebb2
Author: flagdanger <supersags@gmail.com>
Date:   Mon Aug 19 11:31:28 2024 -0400

    PCUObj renamed to PCU, PCU remaned to PCU_C

...
 pcu/{pcu.cc => pcu_c.cc}                              |   2 +-
...
 55 files changed, 297 insertions(+), 300 deletions(-)
```

flagdanger commented:
@cwsmith Can you run tests again? Also, the file level doxygen comment that used to be in pcu.c doesn't exist in the new pcu_c.cc file. How can I put it back into the file and make sure it shows up on the scorec doxygen page? 

There are still unresolved conflicts with the pull request, but I believe the differences are necessary.

cwsmith commented:
/runtests

cwsmith commented:
> Also, the file level doxygen comment that used to be in pcu.c doesn't exist in the new pcu_c.cc file. How can I put it back into the file and make sure it shows up on the scorec doxygen page?

The following should work:
```
git checkout master
cp pcu/pcu.c $HOME/.
git checkout pcu-object-breaking
#copy and paste needed block of text into pcu_c.cc 
git commit pcu_c.cc
```

You can then run 'make doc' (assuming the `doxygen` binary was in a place cmake could find it) and in the `doc` directory (in the build directory) there will be html files you point your web-browser at and view.

jacobmerson commented:
@flagdanger where are we at with the documentation we discussed here: https://github.com/SCOREC/core/pull/433#issuecomment-2452463824

flagdanger commented:
@jacobmerson I can make a draft of documentation at 6 today. Might need review or changes. These conflicts are sticking around after the first merge. I can try again.

jacobmerson commented:
Cameron and I can work on helping with the merge if needed. Make sure to sync your develop branch with SCOREC/develop before the merge

flagdanger commented:
@jacobmerson I tried merging again, it did not make the conflicts go away.

jacobmerson commented:
@flagdanger Cameron and I will work on the merge. Are you set with the documentation changes?

jacobmerson commented:
@cwsmith I think the docs are mostly in order. Take a look and see if there are additional aspects you want. Let's touch base after Thanksgiving to do the final merge.

cwsmith commented:
The documentation looks good.  I reworked the opening description https://github.com/SCOREC/core/pull/433#issue-2336201158 to be more verbose.  I'll resolve the merge conflicts and check the delta wing results.

jacobmerson commented:
Thanks!

cwsmith commented:
@jacobmerson Please take a look at this PR when you get a chance.  https://github.com/jacobmerson/core/pull/4

cwsmith commented:
/runtests

cwsmith commented:
The deltawing test run on piglet (32 procs) was 2.7% slower with this branch (450s) than develop (438s).  Each test was run twice and had consistent performance.  This seems to be more than 'noise', but I think we should proceed with the merge.  

The respective source for the tests are here:
https://github.com/SCOREC/core/tree/cws/developDeltaAdapt @ https://github.com/SCOREC/core/commit/8dc400d85b8862ab2b5f1b7c1707b034f708bc12
https://github.com/SCOREC/core/tree/cws/pcu-object-breaking_deltaAdapt @ https://github.com/SCOREC/core/commit/bedb3c2105a1aa9dcfc48f6fa304bc584925d086

## partition

```mpirun -np 32 ./zsplit /lore/smithc11/projects/deltaWingAdapt/pumiAdapt/deltaWing.dmg  /lore/smithc11/projects/deltaWingAdapt/500kMetric/deltaWing_500kMetric.smb p32/ 32```

## run

```

jacobmerson commented:
My guess would be that we are calling the pcu copy constructor somewhere and it’s enough to cause a mild slow down. I can try to check on this theory. I think we should proceed even if we cannot track that down.

jacobmerson commented:
@cwsmith should `-DCMAKE_BUILD_TYPE` be set to release in the build script?

cwsmith commented:
@jacobmerson By default the pumi build system uses `-O2 -g` ([via these options](https://github.com/SCOREC/core/blob/54b2a9aa91d99ca17f7d2d37446f0629fddf015c/cmake/bob.cmake#L51-L52)).  The compile and link line for the delta wing driver are below.  Do you think `-O3` without debug symbols will make a difference?

```
cd /space/cwsmith/pumi30Release/buildPumiOptonSimonOmegaoff_30/test && /opt/scorec/spack/rhel9/v0201_4/install/linux-rhel9-x86_64/gcc-12.3.0/mpich-4.1.1-xpoyz4tqgfxtrm6m7qq67q4ccp5pnlre/bin/mpicxx -DHAVE_CLOCK_GETTIME -DHAVE_SIMMETRIX -DOMPI_SKIP_MPICXX -DPUMI_HAS_ZOLTAN <...all the include paths> -O2 -g  -std=c++11 -MD -MT test/CMakeFiles/anisoDelta_ma_test.dir/anisoDelta_ma_test.cc.o -MF CMakeFiles/anisoDelta_ma_test.dir/anisoDelta_ma_test.cc.o.d -o C     MakeFiles/anisoDelta_ma_test.dir/anisoDelta_ma_test.cc.o -c /space/cwsmith/pumi30Release/core/test/anisoDelta_ma_test.cc
[100%] Linking CXX executable anisoDelta_ma_test
cd /space/cwsmith/pumi30Release/buildPumiOptonSimonOmegaoff_30/test && /opt/scorec/spack/rhel9/v0201_4/install/linux-rhel9-x86_64/gcc-12.3.0/cmake-3.26.3-2duxfcdiykykfynkadpp2m2y6zrmuvz4/bin/cmake -E cmake_link_script CMakeFiles/anisoDelta_ma_test.dir/link.txt --verbose=1
/opt/scorec/spack/rhel9/v0201_4/install/linux-rhel9-x86_64/gcc-12.3.0/mpich-4.1.1-xpoyz4tqgfxtrm6m7qq67q4ccp5pnlre/bin/mpicxx  -O2 -g  CMakeFiles/anisoDelta_ma_test.dir/anisoDelta_ma_test.cc.o -o anisoDelta_ma_test   <... all the libs>
```
